### DISCUSSION TITLE
Database connection management and SQL execution feature

### DISCUSSION THREAD

**Discussion by sm750s** [open-webui/open-webui#7095] [General]
I would like to request a feature in OpenWebUI that enables the execution of SQL queries provided as responses from the Assistant. OpenWebUI has already added the ability to execute Python code and render HTML, and adding SQL execution capability would further enhance its usability.

To implement this, the following features seem necessary:

- **Database Connection Management**: Administrators should be able to manage database connections within OpenWebUI. Additionally, each model should specify which database connection to use when executing SQL.
- **SQL Execution**: If an Assistant response contains SQL code in a code block, users should be able to click "Run Code." The backend will then use the database connection linked to the model to execute the SQL query. The query results should be displayed as a Markdown table on the screen. (It might also be necessary to manage settings like the maximum number of rows returned.)

I hope this fantastic feature will be made available in OpenWebUI soon.

**Comment by Taehui** [comment]
This is an extremely essential feature. It would be even more useful if various database connectors are supported. I sincerely hope this functionality gets implemented.

**Comment by sm750s** [comment]
I think if the correct table schema is provided to the LLM through the connected Tools or Knowledge, LLM will be able to generate appropriate SQL queries in response to user questions. Additionally, it is expected that users will want to execute the generated SQL queries immediately to check the results.

**Comment by smonux** [comment]
I have in mind doing something similar as a Tool. Instead of launching the sql's by hand, the LLM would do it directly.

 The idea would be to use Apache Drill as a "DB Proxy" (or may be Trino or something similar), so it encapsulates all the connection handling.

I even started coding it (it doesn't work, but you get the idea):

```
"""
title: SQL Data Sources Tool (Apache Drill)

"""

import requests
import json
from pydantic import BaseModel, Field

class Tools:
    class Valves(BaseModel):
        DRILL_URL: str = Field(
            default="http://localhost:8047",
            description="The URL of the Apache Drill instance."
        )
        MAX_ROWS : int = Field(
            default=100, 
            description="Maximum rows returned"
        )

    def __init__(self):
        self.valves = self.Valves()
        self.drill_url = self.valves.DRILL_URL
        self.max_rows = self.valves.MAX_ROWS

    async def get_structure(self):
        """Lists all schemas available in Apache Drill."""
        url = f"{self.drill_url}/storage.json"
        response = requests.get(url)
        if response.status_code == 200:
            return response.json()
        else:
            return f"Error: {response.status_code} - {response.text}"

    async def execute_sql(self, query:str, __event__) -> str:
        """Executes a given SQL query on Apache Drill."""
        url = f"{self.drill_url}/query.json"
        payload = {"queryType": "SQL", "query": query}
        response = requests.post(url, json=payload)

        if response.status_code == 200:
            retval = response.text
        else:
            retval = f"Error: {response.status_code} - {response.text}"

        return retval


    async def explain_plan_sql(self, query:str, __event__) -> str:
        """Executes a given SQL query on Apache Drill."""
        url = f"{self.drill_url}/query.json"
        payload = {"queryType": "SQL",
                   "query": f"explain plan for {query} "}
        response = requests.post(url, json=payload)

        if response.status_code == 200:
            jsretval = response.json()
            # we don't need it (too verbose)
            del jsretval["rows"][0]["json"]
            retval = json.dumps(jsretval)
        else:
            retval = f"Error: {response.status_code} - {response.text}"

        return retval
```

Setting up drill isn't trivial, though. But it supports JDBC sources, at least in theory, so the compatibility range would be pretty good.

**Reply by sm750s** [reply]
The suggestions you provided are excellent ideas. In fact, I have also been trying out executing queries through Tools or Actions. However, I believe it would be great to enhance the user experience by allowing SQL to perform actions such as modifying the code in the returned code block, executing it immediately, and receiving the results.
I envision being able to partially edit the query generated by the LLM, execute it instantly, view the results, and analyze the data in chart form using Mermaid.

**Comment by stockarea** [comment]
Is this feature anywhere near to be implemented?
